### PR TITLE
Adding transphylo, gubbins, bactdating

### PR DESCRIPTION
**Pull request by johannahelene**
Work in progress - still local paths, problems with fastbaps output

- [x] Remove local paths
- [x] Fix fastballs output

### CODE REVIEW

File: Snakefile [hunk 1]
```diff
@@ -32,6 +32,8 @@ rule cluster_summary:
         "output/all_clusters.txt"
     params:
         levels=config['fastbaps']['levels']
+    log:
+        "logs/cluster_summary.log"
```
Comment (johnlees) [inline Snakefile:36]: Is this log file used/does it contain any output?

File: Snakefile [hunk 2]
```diff
@@ -153,21 +155,113 @@ rule iq_tree:
         rooted="output/strains/{strain}/besttree.nwk",
         unrooted=temp("output/strains/{strain}/besttree.unrooted.treefile"),
         iqtree=temp("output/strains/{strain}/besttree.unrooted.iqtree"),
-        ckp=temp("output/strains/{strain}/besttree.unrooted.ckp.gz")
+        ckp=temp("output/strains/{strain}/besttree.unrooted.ckp.gz"),
```
Comment (johnlees) [inline Snakefile]: comma seems incorrect?

File: scripts/run_iqtree.py [hunk 3]
```diff
@@ -37,3 +40,26 @@ def midpoint_root(infile, outfile):
...
+#         # new_align_file.write('\n')
+#     else:
+#         new_align_file.write(line + '\n')
+#         # new_align_file.write('\n')
+# new_align_file.close()
+#
+# # os.remove(snakemake.input.alignment)
+# os.rename(snakemake.params.temp ,snakemake.input.alignment)
+# # delete align_variants
+# # rename temp_align_variants to align_variants
```
Comment (johnlees) [inline scripts/run_iqtree.py]: Remove any commented code that's not being run

File: Snakefile [hunk 4]
```diff
@@ -153,21 +155,113 @@ rule iq_tree:
...
     log:
         "output/strains/{strain}/besttree.unrooted.log"
     params:
         enabled=config['iqtree']['enabled'],
         mode=config['iqtree']['mode'],
         model=config['iqtree']['model'],
-        prefix="output/strains/{strain}/besttree.unrooted"
+        prefix="output/strains/{strain}/besttree.unrooted",
+        temp="output/strains/{strain}/temp_align_variants.aln"
```
Comment (johnlees) [inline Snakefile]: I don't think this is used/needed

File: Snakefile [hunk 5]
```diff
@@ -153,21 +155,113 @@ rule iq_tree:
...
+rule gubbins:
+    input:
+        alignment="output/strains/{strain}/align_variants.aln",
+        start_tree="output/strains/{strain}/besttree.nwk"
+    output:
+        final_tree="output/strains/{strain}/gubbins.final_tree.tre",
+    group:
+        "gubbins"
+    # log:
+    #     "logs/gubbins_{strain}.log"
```
Comment (johnlees) [inline Snakefile]:
```suggestion
```

File: Snakefile [hunk 6]
```diff
@@ -153,21 +155,113 @@ rule iq_tree:
...
+    params:
+        prefix=config['gubbins']['prefix'],
+        tree_builder=config['gubbins']['tree-builder'],
+        min_snp=config['gubbins']['min-snps'],
+        min_window=config['gubbins']['min-window-size'],
+        max_window=config['gubbins']['max-window-size'],
+        iterations=config['gubbins']['iterations'],
```
Comment (johnlees) [inline Snakefile]: All of these need to be passed to the gubbins command below

File: Snakefile [hunk 7]
```diff
@@ -153,21 +155,113 @@ rule iq_tree:
...
+    group:
+        "outbreak"
+    log:
+        "logs/bactdating_{strain}.log"
+    conda:
+        config["poppipe_location"] + "/envs/transphylo.yml"
+    threads:
+        4
+    shell:
+        "Rscript --vanilla scripts/run_bactDating.R output/strains/{wildcards.strain}/gubbins {input.metadata}  {output.sorted} {output.rds} > {log}"
```
Comment (johnlees) [inline Snakefile]: Might want to add `--no-save --no-restore`, I can't remember but worth checking

File: Snakefile [hunk 8]
```diff
@@ -153,21 +155,113 @@ rule iq_tree:
...
+
+rule transmission:
```
Comment (johnlees) [inline Snakefile:225]: Add a comment that this is a dummy target rule

File: Snakefile [hunk 9]
```diff
@@ -153,21 +155,113 @@ rule iq_tree:
...
+        dateT=config['transphylo']['dateT'],
+        gubbins="output/strains/{strain}/gubbins"
+    log:
+        "logs/transphylo_{strain}.log"
+    conda:
+        config["poppipe_location"] + "/envs/transphylo.yml"
+    threads:
+        4
+    shell:
+        "Rscript --vanilla scripts/run_transPhylo.R --rds {input.rds} --sorted {input.sorted} --output {output}\
```
Comment (johnlees) [inline Snakefile:259]: Similarly here, if no save and no restore should be added

File: Snakefile [hunk 10]
```diff
@@ -228,7 +321,7 @@ rule generate_dot:
     conda:
         config["poppipe_location"] + "/envs/mandrake.yml"
     threads:
-        64
+        4
```
Comment (johnlees) [inline Snakefile]: Why did you change all the threads to 4? I think snakemake uses up to the threads specified, so if you have only 4 available it won't try and use 64?

File: config.schema.yml [hunk 11]
```diff
@@ -48,6 +48,14 @@ properties:
...
       - model
 
+  # transphylo:
+  #   type: object
+  #   properties:
+  #     script:
+  #       type: string
+  #   required:
+  #     -
+
```
Comment (johnlees) [inline config.schema.yml]: This file needs to updated for all the new rules

File: config.yml [hunk 12]
```diff
@@ -2,10 +2,11 @@
...
 # Files output by the PopPUNK run
-poppunk_rfile: rlist.txt
-poppunk_clusters: listeria_clusters.csv
-poppunk_h5: listeria.h5
-
+poppunk_rfile: ../PopPUNK/without_hash/rfiles.txt
+poppunk_clusters: ../PopPUNK/without_hash/mass_database/mass_database_clusters.csv
+poppunk_h5: ../PopPUNK/without_hash/mass_database/mass_database.h5
+transmission_metadata: "meta_data_edited_new_strain_name.csv"
+# transmission_metadata: /Users/wachsmannj/Documents/pp/meta_data_edited_new_strain_name.csv
```
Comment (johnlees) [inline config.yml]: Leave all of this section with some defaults that serve as an example

File: config.schema.yml (hunk 11 above)
Comment (johnlees) [reply config.schema.yml]: Base this on your new `config.yml`

File: config.yml [hunk 13]
```diff
@@ -14,6 +15,28 @@ ska:
...
 
+# gubbins options
+gubbins:
+  prefix: gubbins
+  tree-builder: iqtree
+  min-snps: 3
+  min-window-size: 100
+  max-window-size: 10000
+  iterations: 10
+  dir: output/strains/{strain}/
```
Comment (johnlees) [inline config.yml]: I don't think this should be an input parameter, it should just be set in the snakefile

File: Snakefile [hunk 14]
```diff
@@ -153,21 +155,113 @@ rule iq_tree:
...
+    # log:
+    #     "logs/gubbins_{strain}.log"
+    params:
+        prefix=config['gubbins']['prefix'],
+        tree_builder=config['gubbins']['tree-builder'],
+        min_snp=config['gubbins']['min-snps'],
+        min_window=config['gubbins']['min-window-size'],
+        max_window=config['gubbins']['max-window-size'],
+        iterations=config['gubbins']['iterations'],
+        directory=config['gubbins']['dir'],
```
Comment (johnlees) [inline Snakefile]: This should just be set, not read from the config (see other comment on config.yml)

File: config.yml (hunk 13 above)
Comment (johnlees) [reply config.yml]: Also use underscores rather than dashes for consistency

File: config.yml [hunk 15]
```diff
@@ -14,6 +15,28 @@ ska:
...
+
+# transphylo options
+transphylo:
+  w_shape: 10
+  w_scale: 0.1
+  mcmcIterations: 1000
+  startNeg: 0.2739726 #100/365
```
Comment (johnlees) [inline config.yml]: Just use 0.1 as an example default

File: config.yml [hunk 16]
```diff
@@ -14,6 +15,28 @@ ska:
...
+  startOff_r: 1
+  startOff_p: 0.5
+  startPi: 0.5
+  optiStart: 2
+  dateT: 2008.01
```
Comment (johnlees) [inline config.yml]: Why this value?

File: envs/gubbins.yml [hunk 17]
```diff
@@ -0,0 +1,71 @@
+name: gubbins
```
Comment (johnlees) [inline envs/gubbins.yml:1]: This should be minimal, not the full conda list. See the other environments for an example

File: envs/transphylo.yml [hunk 18]
```diff
@@ -0,0 +1,248 @@
+name: transphylo
```
Comment (johnlees) [inline envs/transphylo.yml:1]: Ditto here, should be minimal

File: scripts/number_clusters.py [hunk 19]
```diff
@@ -16,11 +16,14 @@
 
     # Read each fastbaps file
     for fb_file in snakemake.input:
+        print(snakemake.input)
+        print(fb_file)
```
Comment (johnlees) [inline scripts/number_clusters.py]: I think these print statements should be removed?

File: scripts/run_iqtree.py [hunk 20]
```diff
@@ -1,5 +1,7 @@
 import subprocess
 from shutil import copyfile
+from pandas import *
```
Comment (johnlees) [inline scripts/run_iqtree.py]: Can just import `read_table`

File: scripts/run_transPhylo.R [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+  make_option("--startOffr", type="double"),
+  make_option("--startOffp", type="double"),
+  make_option("--startPi", type="double"),
+  make_option("--optiStart", type="double"),
+  make_option("--dateT", type="double")
+)
+
+opt <- parse_args(OptionParser(option_list=option_list))
+
+# TODO: transmission plot
```
Comment (johnlees) [inline scripts/run_transPhylo.R]: Remove this now done?

File: scripts/run_transPhylo.R [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+r_file <- gsub(" ", "", paste("/Users/wachsmannj/Documents/pp/PopPIPE/", opt$rds))
```
Comment (johnlees) [inline scripts/run_transPhylo.R]: Hard-coded path here needs to be fixed

File: Snakefile [hunk 23]
```diff
@@ -153,21 +155,113 @@ rule iq_tree:
...
+    threads:
+        4
+    shell:
+        "pushd {params.directory} &&\
+        run_gubbins.py {params.aln} --prefix {params.prefix} --starting-tree {params.btree} --threads {threads} \
+        && popd"
+
+rule bactdating:
+    input:
+        final_tree="output/strains/{strain}/gubbins.final_tree.tre",
```
Comment (johnlees) [inline Snakefile:208]: You give the tree as input, but then use the whole gubbins directory in the command below? Would be better to use `{input.final_tree}` rather than generating this path twice

File: Snakefile (hunk 2 above)
Comment (johnlees) [reply Snakefile]: Also I think we may just want to remove the checkpoint file, as I'm not sure it's always generated (and it's not used downstream)

**Review by johnlees — CHANGES_REQUESTED** [review]
Also change `version.py` to 1.1.0

File: Snakefile (hunk 1 above)
Comment (johannahelene) [reply Snakefile:36]: oh yes there was a problem with the log file because for gubbins the directory is changes because gubbins creates file in `.` so then I would do:
`log:
        "../../../logs/gubbins_{strain}.log"`
but I am not sure if that is okay?

File: scripts/run_iqtree.py (hunk 3 above)
Comment (johannahelene) [reply scripts/run_iqtree.py]: I wrote this because I think it was transphylo won't work if the isolate names had a `#` in the name. Because then align_variants.aln is changed and it reports an error. Should I un-comment it or just delete it?

File: scripts/number_clusters.py (hunk 19 above)
Comment (johannahelene) [reply scripts/number_clusters.py]: yes

File: Snakefile (hunk 2 above)
Comment (johannahelene) [reply Snakefile]: done!

File: Snakefile (hunk 4 above)
Comment (johannahelene) [reply Snakefile]: depends on if I remove or un-comment the other code from above

File: Snakefile (hunk 23 above)
Comment (johannahelene) [reply Snakefile:208]: I had it that way because gubbins is already looking for `final_tree.tre` so when I give `{input.final_tree}` as an input it does not work because: `cannot open file 'output/strains/10/gubbins.final_tree.tre.final_tree.tre': No such file or directory`

File: config.yml (hunk 13 above)
Comment (johannahelene) [reply config.yml]: oh okay I can take them out. I was under the assumption that we had talked in a meeting to include only a few options but of gubbins but to include `--tree-builder`, all the recombination detection options and `--iterations`. But I can change that

File: Snakefile (hunk 1 above)
Comment (johnlees) [reply Snakefile:36]: I think that would be fine

File: scripts/run_iqtree.py (hunk 3 above)
Comment (johnlees) [reply scripts/run_iqtree.py]: If this fixes the issue with # in sample names then you should uncomment the code.

Looking at the function, it would be better to first check if there are any `#` in the sample names and only run if that's the case, e.g.
```
new_names = []
replacement_needed = False
for name in og_files:
	og_name=name.split()[0]
    if '#' in og_name:
        new_names.append(og_name.replace('#', '_'))
        replacement_needed = True
    else:
    	new_names.append(og_name)
    	
if replacement_needed:
	align_file=open(snakemake.input.alignment, 'r')
	new_align_file = open(snakemake.params.temp, 'w')
	for line in align_file:
		line=line.strip()
		if ">" in line:
	        new_align_file.write(">" + new_names.pop(0) + '\n')
	    else:
	    	new_align_file.write(line + '\n')
	new_align_file.close()
	align_file.close()
	os.rename(snakemake.params.temp ,snakemake.input.alignment)
```

File: Snakefile (hunk 4 above)
Comment (johnlees) [reply Snakefile]: temp is always deleted I think, so should remove

File: config.yml (hunk 13 above)
Comment (johnlees) [reply config.yml]: Sorry, I just meant `dir` should be in the snakefile. The others are all fine, just change `max-window-size` to `max_window_size` etc

File: Snakefile (hunk 23 above)
Comment (johnlees) [reply Snakefile:208]: I mean by changing your `scripts/run_bactDating.R` to use this input rather than the full path, if that works

Comment (johannahelene) [reply Snakefile:208]: I am using the `loadGubbins` function in `scripts/run_bactDating.R` which automatically assumes that the name of the file is `final_tree.tre`but I'll check in the gubbins doc if it is possible to change that

Comment (johnlees) [reply Snakefile:208]: Ah I see, probably not then! In that case just add a comment into the snakefile and/or Rscript explaining this

File: config.yml (hunk 16 above)
Comment (johannahelene) [reply config.yml]: I wanted to set it to `max(dates)` but then I always get this error: `The parameter dateT cannot be smaller than the date of last sample`.
So then I assumed it had to be just a tiny bit larger than that

File: Snakefile (hunk 7 above)
Comment (johannahelene) [reply Snakefile]: I think `--vanilla` should be taking care of that already.
"The `--vanilla` argument makes sure that user-specific R settings are ignored, and that there is no saving or restoring of workspaces. This makes the script more portable to other systems." ([r-bloggers.com](https://www.r-bloggers.com/2019/11/r-scripts-as-command-line-tools/))

File: Snakefile (hunk 9 above)
Comment (johannahelene) [reply Snakefile:259]: same as above

File: Snakefile (hunk 23 above)
Comment (johannahelene) [reply Snakefile:208]: Seems like the [code](https://github.com/xavierdidelot/BactDating/blob/0e93c298ffe3b3006c06d95f40091889a2f6d1e7/R/loadGubbins.R#L7) is wanting only the prefix so then that wouldn't work

File: scripts/run_iqtree.py [hunk 24]
```diff
@@ -24,7 +26,8 @@ def midpoint_root(infile, outfile):
 
 # Change any hashes in names back from underscores
```
Comment (johannahelene) [inline scripts/run_iqtree.py:27]: The other function does not work yet so I took it out

File: Snakefile [hunk 25]
```diff
@@ -246,7 +247,6 @@ rule transphylo:
         startOff_p=config['transphylo']['startOff_p'],
         startPi=config['transphylo']['startPi'],
         optiStart=config['transphylo']['optiStart'],
-        dateT=config['transphylo']['dateT'],
```
Comment (johnlees) [inline Snakefile]: I think you do want to leave this in, and in the rule itself below (sorry if that was unclear), but just set a default of zero. In your R script, then if it's zero set it to max(dates) + a small amount as you originally did.

This is just because the default in the config is a bit too specific, but we still want to let people set it.

File: config.schema.yml [hunk 26]
```diff
@@ -48,13 +50,60 @@ properties:
...
+      default: 10
+  required:
+    - prefix
+    - tree_builder
+    - min_snps
+    - min_window_size
+    - max_window_size
+    - iterations
+
+  transphylo:
```
Comment (johnlees) [inline config.schema.yml:80]: Add dateT in here too (as noted above)

File: config.schema.yml [hunk 27]
```diff
@@ -108,6 +157,7 @@ required:
   - poppunk_rfile
   - poppunk_clusters
   - poppunk_h5
+  - transmission_metadata
```
Comment (johnlees) [inline config.schema.yml]: I think this should actually be optional (so remove from this list)

File: envs/gubbins.yml [hunk 28]
```diff
@@ -5,67 +5,5 @@ channels:
...
-  - tzdata=2022a=h191b570_0
-  - wheel=0.37.1=pyhd8ed1ab_0
-  - xorg-libxau=1.0.9=h35c211d_0
-  - xorg-libxdmcp=1.1.3=h35c211d_0
-  - xz=5.2.5=haf1e3a3_1
-  - zlib=1.2.12=hfe4f2af_1
-  - zstd=1.5.2=ha9df2e0_1
-prefix: /Users/wachsmannj/opt/anaconda3/envs/gubbins
+  - gubbins >= 3.1.0
+  - python =3.9.13
```
Comment (johnlees) [inline envs/gubbins.yml]: Remove python, it's 'special' in conda
```suggestion
  - gubbins >= 3.1.0
```

File: scripts/run_transPhylo.R [hunk 29]
```diff
@@ -13,39 +13,37 @@ option_list <- list(
...
+
   make_option("--wshape", type="double"),
   make_option("--wscale", type="double"),
   make_option("--mcmcIterations", type="integer"),
   make_option("--startNeg", type="double"),
   make_option("--startOffr", type="double"),
   make_option("--startOffp", type="double"),
   make_option("--startPi", type="double"),
-  make_option("--optiStart", type="double"),
-  make_option("--dateT", type="double")
```
Comment (johnlees) [inline scripts/run_transPhylo.R]: Add back in

**Review by johnlees — APPROVED** [review]
